src/riak_kv_entropy_manager.erl

```diff
@@ -644,14 +664,52 @@ maybe_upgrade(State=#state{trees_version = VTrees}) ->
 check_exchanges_and_upgrade(State) ->
     case riak_kv_entropy_info:all_sibling_exchanges_complete() of
         true ->
-            lager:notice("Starting AAE hashtree upgrade"),
-            State#state{pending_version=0};
+            %% Now check nodes who havent reported success in the ETS table
+            {ok, Ring} = riak_core_ring_manager:get_my_ring(),
+            Nodes = riak_core_ring:all_members(Ring),
+            case check_remote_exchanges(Nodes,[]) of
+                complete ->
+                    lager:notice("Starting AAE hashtree upgrade"),
+                    State#state{pending_version=0};
+                _ ->
+                    State
+            end;
         _ ->
             State
     end.
 
+-spec check_remote_exchanges(list(), list()) -> complete | incomplete.
+check_remote_exchanges([], Acc) ->
```

Comment:
Yea - us functional programmers always seem to start with "let's write a recursive list processing function" rather than looking at other options first :)